File: python/2024/10_hoof_it/aoc202410.py
"""AoC 10, 2024: Hoof It."""

# Standard library imports
import collections
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return {
        (row, col): int(char)
        for row, line in enumerate(puzzle_input.split("\n"))
        for col, char in enumerate(line)
    }


def part1(grid):
    """Solve part 1."""
    return sum(
        len(walk(grid, trailhead)) for trailhead, height in grid.items() if height == 0
    )


def part2(grid):
    """Solve part 2."""
    return sum(
        sum(walk(grid, trailhead).values())
        for trailhead, height in grid.items()
        if height == 0
    )


def walk(grid, start, target=9):
    """Walk the grid and count how many paths that can be walked to reach targets."""
    layer = {start: 1}
    for height in range(1, target + 1):
        new_layer = collections.defaultdict(int)
        for pos, num_paths in layer.items():
            for dir in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                new_pos = (pos[0] + dir[0], pos[1] + dir[1])
                if grid.get(new_pos) == height:
                    new_layer[new_pos] += num_paths
        layer, new_layer = new_layer, collections.defaultdict(int)

    return layer


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2024/10_hoof_it/aoc202410.ex
defmodule AOC2024.Day10 do
  @moduledoc """
  Advent of Code 2024, day 10: Hoof It.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    for {line, row} <- puzzle_input |> String.split("\n", trim: true) |> Enum.with_index(),
        {char, col} <- line |> String.split("", trim: true) |> Enum.with_index(),
        into: %{},
        do: {{row, col}, char |> String.to_integer()}
  end

  @doc """
  Solve part 1.
  """
  def part1(grid) do
    Map.filter(grid, fn {_, height} -> height == 0 end)
    |> Map.keys()
    |> Enum.map(fn trailhead -> find_targets(grid, trailhead, 0) |> MapSet.size() end)
    |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2(grid) do
    Map.filter(grid, fn {_, height} -> height == 0 end)
    |> Map.keys()
    |> Enum.map(&count_paths(grid, &1, 0))
    |> Enum.sum()
  end

  def find_targets(_, pos, 9), do: MapSet.new([pos])

  def find_targets(grid, {row, col}, height) do
    [{0, 1}, {1, 0}, {0, -1}, {-1, 0}]
    |> Enum.map(fn {drow, dcol} ->
      new_pos = {row + drow, col + dcol}

      if Map.get(grid, new_pos) == height + 1,
        do: find_targets(grid, new_pos, height + 1),
        else: MapSet.new()
    end)
    |> Enum.reduce(&MapSet.union/2)
  end

  def count_paths(_, _, 9), do: 1

  def count_paths(grid, {row, col}, height) do
    [{0, 1}, {1, 0}, {0, -1}, {-1, 0}]
    |> Enum.map(fn {drow, dcol} ->
      new_pos = {row + drow, col + dcol}
      if Map.get(grid, new_pos) == height + 1, do: count_paths(grid, new_pos, height + 1), else: 0
    end)
    |> Enum.sum()
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
